<a href="https://colab.research.google.com/github/Zuyuf/Machine_Learning_Specialization/blob/main/1-Machine_Learning_Foundation/3%20-%20WEEK%203%20-%20Classification%20Analysing%20Sentiments/2_WEEK_3_Assognment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Import Truicreate and connect to Drive**

In [ ]:
!pip install turicreate

     |████████████████████████████████| 92.0MB 55kB/s 
     |████████████████████████████████| 86.4MB 80kB/s 
     |████████████████████████████████| 327kB 35.7MB/s 
     |████████████████████████████████| 3.4MB 40.0MB/s 
     |████████████████████████████████| 450kB 36.2MB/s 
     |████████████████████████████████| 51kB 4.2MB/s 
     |████████████████████████████████| 3.8MB 29.3MB/s 
  Created wheel for resampy: filename=resampy-0.2.1-cp36-none-any.whl size=320848 sha256=f38c40039e1d62c98ce239a6b965930a894460ea84325fc554d3cf44e9fc49ed
  Stored in directory: /root/.cache/pip/wheels/ff/4f/ed/2e6c676c23efe5394bb40ade50662e90eb46e29b48324c5f9b
  Created wheel for gast: filename=gast-0.2.2-cp36-none-any.whl size=7542 sha256=b2dbfa0f030152e0d569fa951c718bd410c9d66efb17973ea6337a9308d20537
  Stored in directory: /root/.cache/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd
Successfully built resampy gast
ERROR: tensorflow-probability 0.11.0 has requirement gast>=0.3.2, 

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


# Analyze Product Sentiment

In [157]:
import turicreate as tc

# Read product review data

In [158]:
products = tc.SFrame('/content/gdrive/My Drive/Turicreate/Week 3/amazon_baby.sframe')

In [159]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']
count = {'awesome':0, 'great':0, 'fantastic':0, 'amazing':0, 'love':0, 'horrible':0, 'bad':0, 'terrible':0, 'awful':0, 'wow':0, 'hate':0}

In [160]:
def words_count(x):
  counts = {'awesome':0, 'great':0, 'fantastic':0, 'amazing':0, 'love':0, 'horrible':0, 'bad':0, 'terrible':0, 'awful':0, 'wow':0, 'hate':0}
  for s in x.lower().split():
    if s in counts:
      counts[s] += 1
  
  return counts

# Building a sentiment classifier

## Build word count vectors with Given Words "selected_words"

In [161]:
products['word_count'] = tc.text_analytics.count_words(products['review'])

In [162]:
products['selected_word_count'] = products['review'].apply(words_count)

## **Question 1**
Out of the 11 words in selected_words, which one is most used in the reviews in the dataset?

## **Question 2**
Out of the 11 words in selected_words, which one is least used in the reviews in the dataset?

In [163]:
count_all = count

for d in products['selected_word_count']:
  for k in d:
    count_all[k] += d[k]
  
count_all

{'amazing': 1363,
 'awesome': 2090,
 'awful': 383,
 'bad': 3724,
 'fantastic': 932,
 'great': 45206,
 'hate': 1220,
 'horrible': 734,
 'love': 42065,
 'terrible': 748,
 'wow': 144}

**Great** is used the most in this Data set

**Wow** is least used

# Define what is positive and negative sentiment

In [164]:
#ignore all 3*  reviews
products = products[products['rating']!= 3]

In [165]:
#positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4

In [166]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	selected_word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-------------------------------+-----------+
|           word_count          |      selected_word_count      | sentiment |
+-------------------------------+-------------------------------+-----------+
| {'recommend': 1.0, 'highly... | {'awesome': 0, 'great': 0,... |     1     |
| {'quilt': 1.0, 'of': 1.0, ... | {'awesome': 0, 'great': 0,... |     1     |
| {'tool': 1.0, 'clever': 1.... | {'awesome': 0, 'great': 0,... |     1     |
| {'rock': 1.0, 'many': 1.0,... | {'awesome': 0, 'great': 1,... |     1     |
| {'thumb': 1.0, 'or': 1.0, ... | {'awesome': 0, 'great': 1,... |     1     |
| {'for': 1.0, 'barnes': 1.0... | {'awesome': 0, 'great': 0,... |     1     |
| {'right': 1.0, 'because': ... | {'awesome': 0, 'great': 0,... |     1     |
| {'like': 1.0, 'and': 1.0, ... | {'awesome': 0, 'great': 0,... |     1     |
| {'in': 1.0, 'pages': 1.0, ... | {'awesome': 0, 'great': 0,... |     1     |
| {'tracker': 1.0, 'now': 1.... | {'awesome': 0, 'great': 0,... |     1     |
+-------------------------------+-------------------------------+-----------+
[166752 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [167]:
products['sentiment'].show()

Materializing SArray

# Train our sentiment classifier

In [168]:
train_data,test_data = products.random_split(.8,seed=0)

In [169]:
sentiment_model = tc.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 1.474826     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 3.601251     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 4.651818     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 5.334146     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 6.384000     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 10.128027    | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [170]:
selected_words_model = tc.logistic_classifier.create(train_data,target='sentiment', features=['selected_word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.318026     | 0.844299          | 0.842842            |

| 2         | 3        | 0.533614     | 0.844186          | 0.842842            |

| 3         | 4        | 0.744257     | 0.844276          | 0.843142            |

| 4         | 5        | 0.949138     | 0.844269          | 0.843142            |

| 5         | 6        | 1.161271     | 0.844269          | 0.843142            |

| 6         | 7        | 1.374481     | 0.844269          | 0.843142            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

## **Question 3**
Out of the 11 words in selected_words, which one got the most positive weight in the selected_words_model?

## **Question 4**
Out of the 11 words in selected_words, which one got the most negative weight in the selected_words_model?

In [171]:
coefficients = selected_words_model.coefficients.sort('value')
coefficients.print_rows(num_rows=12)

+---------------------+-----------+-------+---------------------+
|         name        |   index   | class |        value        |
+---------------------+-----------+-------+---------------------+
| selected_word_count |  terrible |   1   | -2.0904999848725767 |
| selected_word_count |  horrible |   1   |  -1.996518005587771 |
| selected_word_count |   awful   |   1   | -1.7646995563131873 |
| selected_word_count |    hate   |   1   | -1.4091640627563562 |
| selected_word_count |    bad    |   1   | -0.9858273699287278 |
| selected_word_count |    wow    |   1   | -0.0541450123332186 |
| selected_word_count |   great   |   1   |  0.8839378948979855 |
| selected_word_count | fantastic |   1   |  0.8913030903043403 |
| selected_word_count |  amazing  |   1   |  0.892802422508508  |
| selected_word_count |  awesome  |   1   |  1.0580088887848447 |
|     (intercept)     |    None   |   1   |  1.3672831522929974 |
| selected_word_count |    love   |   1   |  1.3998983430175014 |
+---------

**love** got the most positive weight in the selected_words_model

 i got **terrible** as most negative weight in the selected_words_model

but Answer is **horrible** most negative weight in the selected_words_model

## **Question 6**
Which of the following ranges contains the accuracy of the sentiment_model in the IPython Notebook from lecture on the test_data?

In [172]:
sentiment_model.evaluate(test_data)['accuracy']

0.9176975738650012

**0.9176975738650012** is accuracy of sentiment_model

## **Question 5**
Which of the following ranges contains the accuracy of the selected_words_model on the test_data?

In [173]:
selected_words_model.evaluate(test_data)['accuracy']

0.8431419649291376

**0.8431419649291376** is accuracy of selected_word_model

## **Question 7**
Which of the following ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data?

In [174]:
test_data['predicted_sentiment'] = sentiment_model.predict(test_data, output_type = 'probability')
test_data['predicted_selected_model'] = selected_words_model.predict(test_data, output_type = 'probability')

In [175]:
Baby_Diaper_sentiment_test_data = test_data[test_data['name']== 'Baby Trend Diaper Champ']
Baby_Diaper_sentiment_model = Baby_Diaper_sentiment_test_data.sort('predicted_sentiment', ascending=False)
sentiment_model.evaluate(Baby_Diaper_sentiment_test_data)['accuracy']

0.8157894736842105

In [176]:
Baby_Diaper_selected_test_datal = test_data[test_data['name']== 'Baby Trend Diaper Champ']
Baby_Diaper_selected_model = test_data.sort('predicted_selected_model', ascending=False)
selected_words_model.evaluate(Baby_Diaper_sentiment_test_data)['accuracy']

0.7894736842105263

**0.811 to 0.843**

## **Question 8**
How do you compare the different learned models with the baseline approach where we are just predicting the majority class?

>The model learned using all words performed much better than the other two. The other two approaches performed about the same.

## **Question 9**
Which of the following ranges contains the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’, according to the sentiment_model from the IPython Notebook from lecture?

In [177]:
products['predicted_sentiment'] = sentiment_model.predict(products, output_type = 'probability')
Baby_Diaper_sentiment_products = products[products['name']== 'Baby Trend Diaper Champ']
Baby_Diaper_sentiment_products.sort('predicted_sentiment', ascending=False)[0]['predicted_sentiment']

0.9999999999895941

**0.9 to 1.0**

## **Question 10**
Consider the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the IPython Notebook from lecture. Which of the following ranges contains the predicted_sentiment for this review, if we use the selected_words_model to analyze it?

In [180]:
products['predicted_selected_model'] = selected_words_model.predict(products, output_type = 'probability')
Baby_Diaper_selected_words_products = products[products['name']== 'Baby Trend Diaper Champ']
Baby_Diaper_selected_words_products.sort('predicted_selected_model', ascending=False)[0]['predicted_selected_model']

0.9984234145936199

**0.9984234145936199** i got the answer as 0.9 to 1.0, but its wrong, can't figure out why

acctual answer is 0.7 - 0.8

## **Question 11**
Why is the value of the predicted_sentiment for the most positive review found using the sentiment_model much more positive than the value predicted using the selected_words_model?

the words from selected_words didn't appear 